### PR TITLE
Multiple assertions develop

### PR DESCRIPTION
**Pull request by basz**
Implements https://github.com/ZF-Commons/zfc-rbac/pull/320 for develop branch

- does so without performance loss if not used
- assertion sets may be nested with different conditions
- does so lazily for each condition
  - an AND condition will break early as soon as one assertions fails
  - an OR condition will break early as soon as one assertions succeed
- assertions are cached, so a next call won't require to construct an assertion or AssertionSet.

documentation forthcoming

### CODE REVIEW

**Comment by basz** [comment]
no sure why this PR also show commits already pulled into develop.

review request is from commit e7fd215 and forward

File: src/Role/HierarchicalRole.php [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+ * This software consists of voluntary contributions made by many individuals
+ * and is licensed under the MIT license.
+ */
+
+namespace ZfcRbac\Role;
+
+/**
+ * Simple implementation for a hierarchical role
+ */
+final class HierarchicalRole extends Role implements HierarchicalRoleInterface
```
Comment (svycka) [inline src/Role/HierarchicalRole.php]: maybe if we use `final` then add to others also?

**Comment by DavidHavl** [comment]
Looks great!!

**Comment by basz** [comment]
@svycka thanks. yes, I will do additional changes like php71, remove zf2 support etc... for now could you have a look at the changes from commit e7fd215 and forward?

**Comment by svycka** [comment]
> @svycka thanks. yes, I will do additional changes like php71, remove zf2 support etc... for now could you have a look at the changes from commit e7fd215 and forward?

@basz I already did, but if you could rebase would be a lot easier to review because now it is very hard to see relevant changes.

**Comment by basz** [comment]
I now figured out what goed wrong. this PR https://github.com/ZF-Commons/zfc-rbac/pull/366 isn't merged into develop yet. Guess I should fix that one first and then rebase this one.

**Comment by svycka** [comment]
@basz  or you can remove commits before e7fd215 also should work

File: composer.json [hunk 2]
```diff
@@ -27,15 +27,15 @@
...
+    "php": "^7.0",
+    "zendframework/zend-servicemanager": "^3.0.3",
+    "zendframework/zend-stdlib": "^3.0"
   },
   "require-dev": {
-    "phpunit/phpunit": "~5.0",
-    "squizlabs/php_codesniffer": "^2.0",
-    "doctrine/common": "~2.4"
+    "malukenho/docheader": "^0.1",
+    "phpunit/phpunit": "^6.0",
```
Comment (prolic) [inline composer.json]: let's use v7

File: src/Assertion/AssertionSet.php [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+ * This software consists of voluntary contributions made by many individuals
+ * and is licensed under the MIT license.
+ */
+
+namespace ZfcRbac\Assertion;
+
+use ZfcRbac\Exception;
+use ZfcRbac\Identity\IdentityInterface;
+
+class AssertionSet implements AssertionInterface
```
Comment (prolic) [inline src/Assertion/AssertionSet.php]: if it's implementing an interface, mark as final

File: src/Role/HierarchicalRole.php (hunk 1 above)
Comment (prolic) [reply src/Role/HierarchicalRole.php]: agree, when it's implementing an interface and has no additional public methods, let's mark as final

File: src/Role/Role.php [hunk 4]
```diff
@@ -1,4 +1,6 @@
 <?php
```
[the remark this answers is not in the document]
Comment (prolic) [reply src/Role/Role.php]: I remember there is some git plugin that fixes that, but I forgot the name.

File: src/Role/Role.php [hunk 5]
```diff
@@ -16,68 +18,52 @@
...
+namespace ZfcRbac\Role;
 
 /**
- * @ORM\Entity
- * @ORM\Table(name="permissions")
+ * Simple implementation for a role without hierarchy
+ * and using strings as permissions
  */
-class Permission implements PermissionInterface
+class Role implements RoleInterface
```
Comment (prolic) [inline src/Role/Role.php]: same as above, if it's implementing an interface, mark as final

File: test/Assertion/AssertionSetTest.php [hunk 6]
```diff
@@ -0,0 +1,262 @@
...
+use ZfcRbac\Exception\InvalidArgumentException;
+use ZfcRbac\Identity\IdentityInterface;
+use ZfcRbacTest\Asset\SimpleAssertion;
+
+/**
+ * @covers \ZfcRbac\Assertion\AssertionSet
+ */
+class AssertionSetTest extends TestCase
+{
+    public function testImplementsAssertionInterface()
```
Comment (prolic) [inline test/Assertion/AssertionSetTest.php]: do we require php 7.1 now? if so, let's add `: void`

File: test/Asset/MockRoleWithPermissionMethod.php [hunk 7]
```diff
@@ -11,12 +29,13 @@ public function getPermissions()
...
-    public function getName()
+    public function getName(): string
     {
         return 'role-with-permission-method';
     }
-    public function hasPermission($permission)
+
+    public function hasPermission(string $permission): bool
     {
         return false;
```
[the remark this answers is not in the document]
Comment (prolic) [reply test/Asset/MockRoleWithPermissionMethod.php]: If someone has a problem when using a MockRole from TestAsset in production, he deserves this problem :)

**Comment by prolic** [comment]
Looks good so far, just some minor notes

**Pull request merged**